File: common/gameBoard/src/main/java/yi/component/board/edits/RemoveNodeEdit.java
package yi.component.board.edits;

import yi.models.go.GameModel;
import yi.models.go.GameNode;

import java.util.Objects;

public final class RemoveNodeEdit extends UndoableEdit {

    private final GameNode parentOfNodeToRemove;
    private final GameNode nodeToRemove;

    public RemoveNodeEdit(GameNode nodeToRemove) {
        if (nodeToRemove.isRoot()) {
            throw new IllegalArgumentException("Cannot remove root node");
        }

        this.nodeToRemove = Objects.requireNonNull(nodeToRemove);
        this.parentOfNodeToRemove =  Objects.requireNonNull(nodeToRemove.getParent());
    }

    @Override
    protected boolean rollbackEditNow(GameModel gameModel) {
        if (!parentOfNodeToRemove.isContinuationOf(gameModel.getRootNode())) {
            throw new IllegalStateException("Attempting to re-do but parent node is no " +
                    "longer part of the game tree");
        }

        gameModel.submitNode(parentOfNodeToRemove, nodeToRemove);
        return true;
    }

    @Override
    protected boolean performEditNow(GameModel gameModel) {
        gameModel.removeNodeSubtree(nodeToRemove);
        return true;
    }

    @Override
    protected boolean canRollback() {
        return false;
    }
}

File: models/src/main/kotlin/codes/nibby/yi/models/editor/edit/RemoveNodeEdit.kt
package codes.nibby.yi.models.editor.edit

import codes.nibby.yi.models.GameModel
import codes.nibby.yi.models.GameNode

class RemoveNodeEdit constructor(private val nodeToRemove: GameNode) :
    GameModelEdit {

    private var parentOfNodeToRemove: GameNode? = null

    init {
        check(!nodeToRemove.isRoot()) { "Cannot remove root node" }
    }

    override fun rollbackChanges(model: GameModel) {
        check (parentOfNodeToRemove!!.isContinuationOf(model.getRootNode())) {
            "Attempting to undo but parent is no longer part of the game tree"
        }
        model.editor.addNode(parentOfNodeToRemove!!, nodeToRemove)
    }

    override fun performChanges(model: GameModel): Boolean {
        check(nodeToRemove.isContinuationOf(model.getRootNode())) {
            "Attempting to remove node but the node is no longer part of the game tree"
        }
        parentOfNodeToRemove = nodeToRemove.parent!!
        model.editor.removeNode(nodeToRemove)
        return true
    }

    override fun dispose() {
        nodeToRemove.dispose()
    }
}